pkg/estimator/client/general.go

```diff
@@ -160,13 +160,257 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+	nodes, err := buildModelNodes(cluster)
+	if err != nil {
+		return -1, err
+	}
+	if len(nodes) == 0 {
+		return 0, nil
+	}
+
+	var sets int64
+	for sets < upperBound {
+		if !placeOneSet(oneSetKinds, nodes) {
+			break
+		}
+		sets++
+	}
+	return sets, nil
+}
+
+// placeOneSet attempts to place exactly ONE full set (all kinds with their per-set replica counts)
+// onto the provided working node capacities (in-place)
+// Returns true if successful
+func placeOneSet(orderedKinds []replicaKind, work []modelNode) bool {
+	for _, k := range orderedKinds {
+		remaining := k.count
+		if remaining <= 0 {
+			continue
+		}
+		// first-fit across nodes
+		for n := range work {
+			if remaining <= 0 {
+				break
+			}
+			fit := maxFit(work[n].cap, k.dem)
+			if fit <= 0 {
+				continue
+			}
+			place := fit
+			if place > remaining {
+				place = remaining
+			}
+			consumeMul(work[n].cap, k.dem, place)
+			remaining -= place
+		}
+		if remaining > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// modelNode holds remaining capacity for a given node across all resource types
+type modelNode struct {
+	cap map[corev1.ResourceName]int64
+}
+
+// buildModelNodes constructs identical nodes for each model grade using its Min vector,
+// repeated AllocatableModelings[grade].Count times. Grades are indexed directly.
+func buildModelNodes(cluster *clusterv1alpha1.Cluster) ([]modelNode, error) {
+	if cluster == nil {
+		return nil, fmt.Errorf("nil cluster")
+	}
+	if cluster.Status.ResourceSummary == nil {
+		return nil, fmt.Errorf("resource summary is nil")
+	}
+	spec := cluster.Spec.ResourceModels
+	allocs := cluster.Status.ResourceSummary.AllocatableModelings
+	if len(spec) == 0 {
+		return nil, fmt.Errorf("no resource models defined")
+	}
+
+	// Build capacity template per grade
+	capsByGrade := make(map[uint]map[corev1.ResourceName]int64, len(spec))
+	for _, m := range spec {
+		tmpl := make(map[corev1.ResourceName]int64, len(m.Ranges))
+		for _, r := range m.Ranges {
+			tmpl[r.Name] = quantityAsInt64(r.Min)
+		}
+		capsByGrade[m.Grade] = tmpl
+	}
+
+	// Accumulate counts by grade
+	countByGrade := make(map[uint]int, len(allocs))
+	for _, a := range allocs {
+		if a.Count < 0 {
+			return nil, fmt.Errorf("negative node count for grade %d", a.Grade)
+		}
+		countByGrade[a.Grade] += a.Count
+	}
+
+	// Collect grades and sort, so that order of nodes is
+	grades := make([]int, 0, len(capsByGrade))
+	for g := range capsByGrade {
+		grades = append(grades, int(g)) // #nosec G115: integer overflow conversion uint -> int
+	}
+	sort.Ints(grades)
+
+	// Emit nodes for grades present in both spec & status.
+	var nodes []modelNode
+	for _, grade := range grades {
+		tmpl, cnt := capsByGrade[uint(grade)], countByGrade[uint(grade)] // #nosec G115: integer overflow conversion int -> uint
+		if tmpl == nil || cnt == 0 {
+			continue
+		}
+		for range cnt {
+			capCopy := make(map[corev1.ResourceName]int64, len(tmpl))
+			for k, v := range tmpl {
+				capCopy[k] = v
+			}
```

Comment:
```suggestion
			capCopy := maps.Clone(tmpl)
```
I just realized that now we have a built-in function for cloning a maps since [Go1.21](https://go.dev/doc/go1.21#maps).